pkg/registry/core/shoot/storage/admin_kubeconfig.go

```diff
@@ -117,16 +122,34 @@ func (r *AdminKubeconfigREST) Create(ctx context.Context, name string, obj runti
 		return nil, errors.NewInvalid(gvk.GroupKind(), shoot.Name, field.ErrorList{fieldErr})
 	}
 
-	resourceDataList := gardencorev1alpha1helper.GardenerResourceDataList(shootState.Spec.Gardener)
+	var (
+		resourceDataList = gardencorev1alpha1helper.GardenerResourceDataList(shootState.Spec.Gardener)
+		ca               *gardencorev1alpha1.GardenerResourceData
+	)
 
-	ca := resourceDataList.Get(v1beta1constants.SecretNameCAClient)
+	ca, err = getCAFromResourceDataList(resourceDataList, nameCAClientReq)
+	if err != nil {
+		return nil, errors.NewInternalError(fmt.Errorf("could not find CA certificate for %s: %w", nameCAClientReq, err))
+	}
+
+	// fall back to cluster CA since not all clusters might have a client CA yet
 	if ca == nil {
-		ca = resourceDataList.Get(v1beta1constants.SecretNameCACluster)
-		if ca == nil {
-			return nil, errors.NewInternalError(fmt.Errorf("certificate authority not yet provisioned"))
+		ca, err = getCAFromResourceDataList(resourceDataList, nameCAClusterReq)
```

Comment:
Q- Why are we trying to get cluster ca at two places? and line 146

Reply:
Here we try to get it via the `labels`, in L146 we get it via the `name` as fall-back for backwards-compatibility (the `ShootState` for some clusters might not yet contain the labels). The L146-related code will be dropped in the next Gardener version.

Reply:
Thanks, I guessed since we are already one version ahead every cluster's `ShootState` must have updated.